convert2rhel/actions/pre_ponr_changes/backup_system.py

```diff
@@ -63,3 +67,96 @@ def run(self):
 
         repo.backup_yum_repos()
         repo.backup_varsdir()
+
+
+class BackupPackageFiles(actions.Action):
+    id = "BACKUP_PACKAGE_FILES"
+
+    def run(self):
+        """Backup changed package files"""
+        loggerinst.task("Prepare: Backup package files")
+
+        super(BackupPackageFiles, self).run()
+
+        package_files_changes = self._get_rpm_va()
+
+        for file in package_files_changes:
+            if file["status"] == "missing":
+                pass  # no action needed now
+            elif re.search(r"[5]", file["status"]):
+                # If the MD5 checksum differs, the content of the file differs
+                restorable_file = backup.NewRestorableFile(file["path"])
+                backup.backup_control.push(restorable_file)
+
+    def _get_rpm_va(self):
+        """Get the output from rpm -Va command from generated during resolving system info
+        to get changes made to package files.
+        Return them as a list of dict, for example:
+        [{"status":"S5T", "file_type":"c", "path":"/etc/yum.repos.d/CentOS-Linux-AppStream.repo"}]
+        """
+        path = os.path.join(LOG_DIR, PRE_RPM_VA_LOG_FILENAME)
+        with open(path, "r") as f:
+            output = f.read()
+
+        return self._parse(output)
+
+    def _parse(self, input):
+        """Parse the output from input"""
+        input = input.strip()
+        lines = input.split("\n")
+        data = []
+
+        for line in lines:
+            parsed_line = self._parse_line(line.strip())
+            if parsed_line:
+                data.append(parsed_line)
+
+        return data
+
+    def _parse_line(self, line):
+        """Return {"status":"S5T", "file_type":"c", "path":"/etc/yum.repos.d/CentOS-Linux-AppStream.repo"}"""
+
+        # Regex explanation:
+        # Match missing or SM5DLUGTP (letters can be replaced by dots or ?):
+        #   (missing|([S\.\?][M\.\?][5\.\?][D\.\?][L\.\?][U\.\?][G\.\?][T\.\?][P\.\?]))
+        # Match whitespace: \s+
+        # Match type of file, can be replaced by any whitecharacter:
+        #   [cdlr\s+]
+        # Match unix path:
+        #   [\/\\](?:(?!\.\s+)\S)+(\.)?
+        regex = r"^(missing|([S\.\?][M\.\?][5\.\?][D\.\?][L\.\?][U\.\?][G\.\?][T\.\?][P\.\?]))\s+[cdlr\s+]\s+[\/\\](?:(?!\.\s+)\S)+(\.)?$"
+
+        ret = re.match(regex, line)
+
+        if not ret:  # line not matching the regex
+            if line.strip() != "":
+                # Line is not empty string
+                loggerinst.debug("Skipping invalid output %s" % line)
+            return
```

Comment:
I'm wondering if we should return a dictionary here with some default values to avoid the ` Explicit returns mixed with implicit (fall through) returns` from codeql. Seems to me that it would be a benefit to have the same return type in this case as well, but with default values, like, `None`.

Reply:
We can filter it out in the function above that depends on this result to only append data that has something other than the default value we don't care.

Reply:
Makes sense, it will be more systematic